## Домашнее задание к занятию 1: "Алгоритм линейной регрессии. Градиентный спуск"

#### Задание:
1. Подберите скорость обучения (alpha) и количество итераций для градиентного спуска.
2. (опция). В этом коде мы избавляемся от итераций по весам, но тут есть ошибка, исправьте ее:

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import datetime as dtm

In [2]:
def calc_mse(y, y_pred):
    err = np.mean((y - y_pred)**2)
    return err

In [3]:
y = [45, 55, 50, 59, 65, 35, 75, 80, 50, 60]
X = np.array([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
              [1, 1, 2, 1, 3, 0, 5, 10, 1, 2]])

In [4]:
W = np.dot(np.dot(np.linalg.inv(np.dot(X, X.T)), X), y)
W

array([47.23214286,  3.91071429])

In [5]:
n = X.shape[1]
alpha = 1e-8
W = np.array([1, 0.5])
W, alpha

(array([1. , 0.5]), 1e-08)

In [6]:
## Исходный код
for i in range(10000):
    y_pred = np.dot(W, X)
    err = calc_mse(y, y_pred)
    for ii in range(W.shape[0]):
        W[ii] -= alpha * (2/n * np.sum(X[ii] * (y_pred - y)))
#         print(X[ii])
#     W -= (alpha * (1/n * 2 * np.sum(X * (y_pred - y))))
    if i % 1000 == 0:
        print(i, W, err)

0 [1.0000011 0.5000034] 3173.15
1000 [1.001103   0.50340287] 3171.872938282029
2000 [1.0022047 0.5068013] 3170.5966340026016
3000 [1.00330621 0.51019867] 3169.321086705227
4000 [1.00440751 0.513595  ] 3168.046295933692
5000 [1.00550862 0.51699028] 3166.7722612320567
6000 [1.00660952 0.52038451] 3165.498982144659
7000 [1.00771023 0.52377769] 3164.2264582161074
8000 [1.00881074 0.52716982] 3162.954688991288
9000 [1.00991106 0.5305609 ] 3161.6836740153612


In [7]:
n = X.shape[1]
alpha = 1e-8
W = np.array([1, 0.5])
W, alpha

(array([1. , 0.5]), 1e-08)

In [8]:
## Без итерации по весам
for i in range(10000):
    y_pred = np.dot(W, X)
    err = calc_mse(y, y_pred)
    W -= alpha * (2/n * np.sum((X * (y_pred - y)), axis=1)) 
    if i % 1000 == 0:
        print(i, W, err)

0 [1.0000011 0.5000034] 3173.15
1000 [1.001103   0.50340287] 3171.872938282029
2000 [1.0022047 0.5068013] 3170.5966340026016
3000 [1.00330621 0.51019867] 3169.321086705227
4000 [1.00440751 0.513595  ] 3168.046295933692
5000 [1.00550862 0.51699028] 3166.7722612320567
6000 [1.00660952 0.52038451] 3165.498982144659
7000 [1.00771023 0.52377769] 3164.2264582161074
8000 [1.00881074 0.52716982] 3162.954688991288
9000 [1.00991106 0.5305609 ] 3161.6836740153612


#### Найдем необходимое количество итераций для разных значений скорости сходимости

In [9]:
n = X.shape[1]
alpha_array = [1e-2, 1e-3, 1e-4, 1e-5, 1e-6]

In [10]:
%%time
eps = 1e-10
num_of_iter = 100000000
W_array = []
iter_array = []
error_array = []

for alpha in alpha_array:
    W = np.array([1, 0.5])
    err = 1e10
    date_start = dtm.datetime.now()
    for iteration in range(num_of_iter):
        y_pred = np.dot(W, X)
        new_err = calc_mse(y, y_pred)
        if abs(new_err - err) < eps:
            W_array.append(W)
            iter_array.append(iteration)
            error_array.append(err)
            print('Скорость сходимости:', alpha) 
            print(' - количество итераций:', iteration)
            print(' - вектор весов:', W)
            print(' - ошибка:', new_err)
            print(' - время:', dtm.datetime.now() - date_start)
            print()
            break
        W -= alpha * (2/n * np.sum((X * (y_pred - y)), axis=1)) 
        err = new_err

Скорость сходимости: 0.01
 - количество итераций: 1250
 - вектор весов: [47.23204948  3.91073153]
 - ошибка: 45.9375000046876
 - время: 0:00:00.069813

Скорость сходимости: 0.001
 - количество итераций: 11441
 - вектор весов: [47.2318441   3.91076945]
 - ошибка: 45.9375000479849
 - время: 0:00:00.407362

Скорость сходимости: 0.0001
 - количество итераций: 103382
 - вектор весов: [47.23119716  3.91088892]
 - ошибка: 45.93750048081665
 - время: 0:00:03.734011

Скорость сходимости: 1e-05
 - количество итераций: 923138
 - вектор весов: [47.2291521   3.91126655]
 - ошибка: 45.93750480880125
 - время: 0:00:30.707945

Скорость сходимости: 1e-06
 - количество итераций: 8124027
 - вектор весов: [47.22268376  3.91246098]
 - ошибка: 45.937548103086506
 - время: 0:04:58.259663

Wall time: 5min 33s


#### При сравнении полученных векторов весов для разных скоростей сходимости с вычисленным теоретическим значением наиболее близким при заданном начальном приближении оказался вектор весов, полученный при следующих параметрах:

    - скорость сходимости = 0.01
    
    - количество итераций = 1250
    
#### При скорости сходимости 0.01 мы получаем минимальную ошибку для выбранного приближения и наиболее быструю сходимость с минимальным количеством итераций. 